SalesforceCommerceCloud/pwa-kit
# @W-13942827 - Reduced Javascript Heap Size Ramp Up

# Description

This PR is **NOT** a fix for the heap size issue which eventually leads to the dev server crashing, but it can help prevent it from happening as often. Mainly what we have here are a couple fixes that **_decrease_** how fast heap size grows. I've seen an decrease in code size increase per save leading to a 2x increase in saves before crashing.

_NOTE: Before I get into what I did, I'll mention that the heap size increase doesn't have anything to do with `webpack-hot-middleware`, this same issue happens if we disabled hot module loading._

## Ok so what does this PR do?

1. **We only enabled server source-maps when we are actually debugging (aka the _inspect_ flag is set).**

Why does this help? Well it's because those `eval-source-maps` end up in memory, and the base issue of the dev server not releasing its references means we have duplicate source maps in memory and they are big (like 30 or 200 mb big). So this means that during normal development we aren't creating those source maps in memory on the server.

2. **Lets not include 3rd party lib source-maps!**

Currently we were including 3rd party library source maps if they exist, adding the `source-map-loader` plugin insures that 3rd party source-maps abide by the same "devtool" setting defined in our configs. 

3. **Fix possible memory leak.**

During heap analysis I saw a lot of references to `jwt-decode` in the commerce-sdk-react lib. I've upgraded the version to the latest major version, I _think_ this helped as well.

## What is the true source of the memory leak?

Our development environment uses `webpack-hot-server-middleware' in order to allow changes to code to be observed in the browser after refreshing your page without having the restart the dev server, furthermore we don't event need to us refresh the browser as `webpack-hot-middleware` does that for us. But that isn't important to our findings here. 

How does `webpack-hot-server-middleware` work? well it will detect changes in your files and hook into the compilation step of webpack. Internally this plugin uses a package called `require-from-string` to compile a module from a "code" string. In in fact is where the issue arises, this package uses node's module object to compile it's code (using a private function none-the-less). I cannot say for certain, but internally in the module object there a reference persists to this code string and never lets it be collected by the garbage collector. Hence we end up having multiples of the webpage runtime (the app) in memory. Once we reach the limit (defaults to 4GB on mac machines) you'll crash.

[Here](https://github.com/floatdrop/require-from-string/issues/27) is a link to a bug that I filed on the `require-from-string` project describing the issue and providing a small node application that highlights the issue. Hopefully this developer will be able to shed some light on the issue, but I believe that issues lies in Node itself. 

If you'd like to test this script out create a node script with the content below:

```
#!/usr/bin/env node

var requireFromString = require('require-from-string');
var cycles = 100000;
var delay = 10;

(function step() {
  requireFromString(`/* ${'x'.repeat(10*512*1024)} */`, '/fake/path/to/file.js');

  if (--cycles > 0) {
    setTimeout(step, delay);
  }
}());
```

And run it like this: 

`node require-string-crash.js`

Next wait for the application to crash. This should only take about 10 seconds.

# Callouts. 

- @kevinxh Pointed out that we could look into clearing the require cache as a way to clear up that memory usage. I in fact looked into this after he mentioned it, but the way `require-from-string` works, I believe that no cache entry is ever made for the compiled code as it is manually using the `_compile` function. I confirmed that the code string isn't located in the cache by printing the contents and specifically looking for the entry using the key (file name) used. 

# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`


# Changes

- Only include source maps in the server when using "inspect", doesn't make sense otherwise to include them if they are never going to be used. 90% of the time people are debugging client code, not on the server.
- Add `source-map-loader` plugin.
- Update some dependencies that looked suspect.

# How to Test-Drive This PR

_This is going to be trick to test, to do this we'll compare memory usage with the `develop` branch. Follow the below steps on both `develop` and `fix-source-maps-2` branch recording your results for comparison at the end. _

- `npm checkout BRANCH_NAME`
- `git clean -dxf && npm ci` // Get a clean state and install dependencies.
- `cd packages/template-retail-react-app`
- `npm start`
- Now open the "Activity Monitor" app and filter on "node". There should be 4 node apps, the largest (by memory is the one that we are interested in).
- Once the application is running in your browser and the first build is complete, record the memory size of the above node application.
- Now in your code editor save a template file (I used pages/home/index.jsx) about 10 times.
- Again record the memory usage.
- Once you have done this to both branches, compare the values your recorded (I recorded a 1.49 GB vs 1.05 GB with this fix).

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
